### PR TITLE
[feature] Allow to specify timeout when requesting features

### PR DESCRIPTION
**Pull request by m-kuhn**
This is potentially interesting for race conditions, but whenever it is specified,
iterators will need to be carefully checked for validity and error handling needs
to be taken care of.

**Disclaimer**
This is not really tested or required, but it would be a pity to just loose this code in a hidden branch.

### CODE REVIEW

File: src/core/qgsvectorlayerfeatureiterator.cpp [hunk 1]
```diff
@@ -211,12 +211,8 @@ QgsVectorLayerFeatureIterator::QgsVectorLayerFeatureIterator( QgsVectorLayerFeat
   if ( mSource->mHasEditBuffer )
   {
     mChangedFeaturesRequest = mProviderRequest;
-    QgsFeatureIds changedIds;
-    QgsChangedAttributesMap::const_iterator attIt = mSource->mChangedAttributeValues.constBegin();
-    for ( ; attIt != mSource->mChangedAttributeValues.constEnd(); ++attIt )
-    {
-      changedIds << attIt.key();
-    }
+    const QgsFeatureIds changedIds = mSource->mChangedAttributeValues.keys().toSet();
```
Comment (nyalldawson) [inline src/core/qgsvectorlayerfeatureiterator.cpp]: See earlier PR comment regarding this - the more verbose code is more efficient, due to https://github.com/KDE/clazy/blob/master/src/checks/level0/README-container-anti-pattern.md

**Comment by m-kuhn** [comment]
I don't know what `timeout = 0` means, Qt doc is quiet on this topic http://doc.qt.io/qt-4.8/qsemaphore.html#tryAcquire-2

**Comment by nyalldawson** [comment]
> I don't know what timeout = 0 means

I was mostly wondering if it would instant fail and could be used for testing. Anyway, looks good to me to merge!

**Comment by nyalldawson** [comment]
Actually one last little thing - I think it's worth putting a note in the dox that not all underlying data providers support the timeout.

**Comment by m-kuhn** [comment]
yes, might be worth a log message as well.

**Comment by m-kuhn** [comment]
oops, no virtual base function to provide a default implementation for... just added to the api.

**Pull request merged**